### PR TITLE
is_mutable_string takes one argument

### LINKED ISSUE #242: is_mutable_string takes one argument
`is_mutable_string` should use a private exemplar of _InputIterator_ instead of requiring the type `I`.

### PR DESCRIPTION
**Pull request by alandefreitas**
fix #242

### CODE REVIEW

File: include/boost/url/grammar/detail/type_traits.hpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        input_it_ex const&)
+    {
+        return false;
+    }
+
+    // 3. Expressions
+    pointer
+    operator->()
+    {
+        return &(operator*());
```
Comment (vinniefalco) [inline include/boost/url/grammar/detail/type_traits.hpp:88]: leave out all the definitions please

File: include/boost/url/grammar/detail/type_traits.hpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_GRAMMAR_DETAIL_TYPE_TRAITS_HPP
+#define BOOST_URL_GRAMMAR_DETAIL_TYPE_TRAITS_HPP
+
```
Comment (vinniefalco) [inline include/boost/url/grammar/detail/type_traits.hpp:12]: missing `<iterator>` and `<cstddef>`

**Comment by vinniefalco** [comment]
You could add a test, ifdef for C++20 that checks against the concept. The test would go in `test/unit/grammar/type_traits.cpp`

**Comment by alandefreitas** [comment]
Done

File: include/boost/url/grammar/detail/type_traits.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
+//
+// Copyright (c) 2016-2019 Vinnie Falco (vinnie dot falco at gmail dot com)
```
Comment (vinniefalco) [inline include/boost/url/grammar/detail/type_traits.hpp:2]: why is my name on this?

File: include/boost/url/grammar/detail/type_traits.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    // 1.1 Type defs
+    using value_type = char;
+    using difference_type = std::ptrdiff_t;
+    using reference = char&;
+    using pointer = char*;
+    using iterator_category = std::input_iterator_tag;
+
+    // 1.2 CopyConstructible
+    input_it_ex(input_it_ex const&) = default;
```
Comment (vinniefalco) [inline include/boost/url/grammar/detail/type_traits.hpp:34]: technically, `=default` counts as a definition :)

File: include/boost/url/grammar/type_traits.hpp [hunk 5]
```diff
@@ -27,24 +28,24 @@ namespace grammar {
 template<class T, class I>
 using is_mutable_string = __see_below__;
 #else
-template<class T, class I, class = void>
+template<class T, class = void>
```
Comment (vinniefalco) [inline include/boost/url/grammar/type_traits.hpp:31]: What about the javadoc? It still says `I`

File: test/unit/grammar/type_traits.cpp [hunk 6]
```diff
@@ -9,3 +9,32 @@
...
+namespace urls {
+namespace grammar {
+
+struct type_traits_test
+{
+    void
+    run()
+    {
+#ifdef __cpp_lib_concepts
+        BOOST_TEST(std::input_iterator<detail::input_it_ex>);
```
Comment (vinniefalco) [inline test/unit/grammar/type_traits.cpp:27]: Nice :)

File: test/unit/grammar/type_traits.cpp [hunk 7]
```diff
@@ -9,3 +9,32 @@
...
+#endif
+        BOOST_TEST(is_mutable_string<std::string>::value);
+        BOOST_TEST_NOT(is_mutable_string<int>::value);
```
Comment (vinniefalco) [inline test/unit/grammar/type_traits.cpp:30]: c'mon man give it something to chew on, maybe `vector<char>`? or how about a `std::string const`? Or how about `std::basic_string<char const, std::char_traits<char>>` ?

**Pull request closed without merging**